libs/wingc/src/type_check/fqn.rs

```diff
@@ -0,0 +1,161 @@
+/// where there may be zero or more namespaces.
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct FQN<'a>(&'a str);
+
+impl<'a> From<&'a str> for FQN<'a> {
+	fn from(fqn: &'a str) -> Self {
+		if fqn.split('.').count() < 2 {
+			panic!("Invalid FQN: {}", fqn);
+		}
+		FQN(fqn)
+	}
+}
+
+impl<'a> From<&'a String> for FQN<'a> {
+	fn from(fqn: &'a String) -> Self {
+		if fqn.split('.').count() < 2 {
+			panic!("Invalid FQN: {}", fqn);
+		}
+		FQN(fqn)
+	}
+}
+
+impl<'a> FQN<'a> {
+	/// Returns the FQN as a string.
+	pub fn as_str(&self) -> &str {
+		self.0
+	}
+
+	/// Returns the FQN as a string without the type name.
+	pub fn as_str_without_type_name(&self) -> &str {
+		let index = self.0.rfind('.').unwrap();
+		&self.0[..index]
+	}
+
+	/// Returns the "assembly" part of the FQN. This is the name of the
+	/// npm library or wing library the type is defined in.
+	#[allow(dead_code)]
+	pub fn assembly(&self) -> &str {
+		self.0.split('.').next().unwrap()
+	}
+
+	/// Returns the "type name" part of the FQN. This is the name of the
+	/// type itself.
+	pub fn type_name(&self) -> &str {
+		self.0.split('.').last().unwrap()
+	}
+
+	/// Returns an iterator over the namespaces of the FQN. The namespaces
+	/// are the parts of the FQN between the assembly name and the type name
+	/// which are used to group types together.
+	pub fn namespaces(&self) -> impl Iterator<Item = &str> {
+		let mut parts = self.0.split('.');
+		parts.next();
+		parts.next_back().unwrap();
+		parts
+	}
+
+	/// Returns true if the FQN represents a "construct base class".
+	///
+	/// TODO: this is a temporary hack until we support interfaces.
+	pub fn is_construct_base(&self) -> bool {
+		// We treat both CONSTRUCT_BASE and WINGSDK_RESOURCE, as base constructs because in wingsdk we currently have stuff directly derived
+		// from `construct.Construct` and stuff derived `cloud.Resource` (which itself is derived from `constructs.Construct`).
+		// But since we don't support interfaces yet we can't import `core.Resource` so we just treat it as a base class.
+		// I'm also not sure we should ever import `core.Resource` because we might want to keep its internals hidden to the user:
+		// after all it's an abstract class representing our `resource` primitive. See https://github.com/winglang/wing/issues/261.
+		self.0 == &format!("{}.{}", WINGSDK_ASSEMBLY_NAME, WINGSDK_RESOURCE) || self.0 == CONSTRUCT_BASE
+	}
+
+	/// Returns true if the FQN belongs to a namespace that is a prefix of the given namespace filter.
+	/// For example, if the FQN is `my_lib.ns1.ns2.MyResource` then the resource is part of these filters:
+	/// - `[]`
+	/// - `["ns1"]`
+	/// - `["ns1", "ns2"]`
+	///
+	/// But it is not part of these filters:
+	/// - `["ns1", "ns2", "ns3"]`
+	/// - `["ns2"]`
+	pub fn is_in_namespace<T: AsRef<str>>(&self, namespace_filter: &[T]) -> bool {
+		self.namespaces().count() >= namespace_filter.len()
+			&& self
+				.namespaces()
```

Comment:
Note that calling `namespaces()` here multiple times is inefficient. It might hint that we should store the FQN as a vector and not call `spilt` each time we want some info out of it. Not critical at this point, but just saying...

Reply:
It's a fair point. I think for now this is cheaper in the respect that most places in the jsii importer would need to do a clone if FQN took ownership of the string, but I can see the argument for both sides... seems like a classic memory/speed tradeoff. :-)